ofrak_core/ofrak/gui/script_builder.py

```diff
@@ -0,0 +1,297 @@
+                children = await parent.get_children(
+                    r_filter=ResourceFilter(
+                        resource.get_most_specific_tags(),
+                        attribute_filters=ResourceAttributeValueFilter(
+                            attribute=attribute,
+                            value=attribute_value)
+                        )
+                    )
+                if len(children) > 1:
+                    attribute_collisions[attribute.__name__] = attribute_value
+                    continue
+                return attribute, attribute_value
+        if len(attribute_collisions) == 0:
+            raise SelectableAttributesError(
+                f"Resource with ID {resource.get_id()} does not have a selectable attribute."
+            )
+        else:
+            msg = []
+            for collision, value in attribute_collisions.items(): 
+                msg.append(f"Resource with ID {resource.get_id()} cannot be uniquely identified by attribute {attribute.__name__} (resource has value {attribute_value}).")
+            raise SelectableAttributesError("\n".join(msg))
+
+    async def _generate_name(self, resource: Resource) -> str:
+        root_resource = await self._get_root_resource(resource)
+        most_specific_tag = list(resource.get_most_specific_tags())[0].__name__.lower()
+        _, selectable_attribute_value = await self._get_selectable_attribute(resource)
+        name = f"{most_specific_tag}_{selectable_attribute_value}"
+        name = re.sub(r"[\s\-\.\/\\]", "_", name)
+        if name in self.script_sessions[root_resource.get_id()].resource_variable_names.values():
+            parent = await resource.get_parent()
+            return f"{self.script_sessions[root_resource.get_id()].resource_variable_names[parent.get_id()]}_{name}"
+        return name
+
+    def _get_session(self, resource_id: bytes) -> ScriptSession:
+        session = self.script_sessions.get(resource_id, None)
+        if session is None:
+            session = ScriptSession()
+            self.script_sessions[resource_id] = session
+
+        return session
+
+    def _get_script(self, resource_id: bytes, target_type: Optional[ActionType] = None) -> str:
+        script = []
+        script.append(self.script_sessions[resource_id].boilerplate_header)
+        for script_action in self.script_sessions[resource_id].actions:
+            # Always include UNDEF actions like variable assignments
+            if (
+                target_type is None
+                or script_action.action_type == target_type
+                or script_action.action_type == ActionType.UNDEF
+            ):
+                script.append(f"{script_action.action}")
+        script.append(self.script_sessions[resource_id].boilerplate_footer)
+        script = "\n".join(script)
+        script = self._dedent(script)
+        res = format_str("\n".join(script), mode=FileMode())
+        script = res.split("\n")
```

Comment:
What is the benefit of splitting the script into lines? I don't really see one, and if there isn't one, it would obsolete and obviate error-prone code in several places if we just passed around a string.

Reply:
The primary benefit is that it simplifies line numbering on the frontend. Rather than having to parse the script on the frontend to increment the line number each time there's a newline, line numbering is easily handled by setting the current line number to the index in the list plus one.

Reply:
I'm not sure that is worth passing everything around as a list of lines, since the string can just as easily be split on the front end. But I also think at this point it doesn't need to be changed.